### PR TITLE
fix bug: right anti join with filter

### LINKED ISSUE #4247: RightSemi/RightAnti HashJoin has bug, the left_indices is never populated, causing failure to apply join filters.
**Describe the bug**

**To Reproduce**

**Expected behavior**

**Additional context**

### PR DESCRIPTION
**Pull request by liukun4515**
# Which issue does this PR close?

Closes #4247

- Fix the bug `right anti` join with filter

- Add ut for this case.

TODO: refactor the hash and combine all join type to the same operation, and reduce the checker of `match`.

# Rationale for this change

# What changes are included in this PR?

# Are these changes tested?

# Are there any user-facing changes?

### CODE REVIEW

File: datafusion/core/src/physical_plan/joins/hash_join.rs [hunk 1]
```diff
@@ -884,32 +904,27 @@ fn build_join_indexes(
...
+                            i as usize,
+                            row,
+                            &left_join_values,
+                            &keys_values,
+                            *null_equals_null,
+                        )? {
+                            left_indices.append(i);
+                            right_indices.append(row as u32);
                         }
                     }
```
Comment (mingmwang) [inline datafusion/core/src/physical_plan/joins/hash_join.rs:927]: Nice! This makes the right anti join logic much more simpler and more consistent with the other join types !

Comment (Dandandan) [reply datafusion/core/src/physical_plan/joins/hash_join.rs:927]: Can't it be combined with the other cases like `JoinType::Inner` (code is the same now?)

Comment (liukun4515) [reply datafusion/core/src/physical_plan/joins/hash_join.rs:927]: > Can't it be combined with the other cases like `JoinType::Inner` (code is the same now?)

yes, I am working on this and refactor the code and combine all same logic together.

So we can hold this pr and wait for the next refactored pr

**Comment by liukun4515** [comment]
convert this pr to draft, and file a new pr to refactor the hash join.
The new pr can also resolve this issue

**Pull request closed without merging**